Fixing CI + Python Bindings (Pkgs 16/N)

padentomasello opened the pull request:
**IMPORTANT: Please do not create a Pull Request without creating an issue first.** Changes *must* be discussed.

**Original Issue**: [corresponding issue on Github]

*Note:* You can add `closes #[issue number]` to automatically close the issue that this PR resolves when it is merged.

### Summary
[Explain the details for this change and the problem that the pull request solves]

### Test Plan (required)
[steps by which you tested that your fix resolves the issue. These might include specific commands and configurations]

Commit:
Vision / Runtime pkg

Commit:
Speech pkg

Commit:
Fix vision tests

Commit:
Add text pkg

Commit:
Move amp to runtime pkg

Commit:
Move plugin to runtime

Commit:
Move halide to pkg

Commit:
If it's common, probably shouldn't be in ext...

Commit:
Split libs. Remove FL -> lib dependency

Commit:
Update namespace app::asr -> pkg::speech

Commit:
app::imgclass / app::objdet --> pkg::vision

Commit:
namespace pkg::runtime

Commit:
namespace pkg::text

Commit:
fl::app::image -> fl::pkg::vision

Commit:
namespace ext --> fl

Commit:
Fixing CI

Commit:
Add pkg/test/CMakeLists.txt

Commit:
PR cleanup

Commit:
Fix CMakeLists verision

Commit:
Cmake -> 3.10

Commit:
Cmake -> 3.10

Commit:
PR Fixes

Commit:
Cmake -> 3.10

Commit:
PR fixes

Commit:
Remove extra line

Commit:
PR Fixes

Commit:
PR fixes

jacobkahn commented on cmake/TestUtils.cmake:
```diff
@@ -48,6 +48,7 @@ function(build_test)
     PUBLIC
     ${GTEST_IMPORTED_TARGETS}
     ${build_test_LIBS}
+    pthread
```
Not portable/doesn't work for libomp with MKL - see what we do [here](https://github.com/flashlight/flashlight/blob/c82d4d05e43c21702d57af70c8071282080da1e2/flashlight/fl/common/CMakeLists.txt#L34)

We should also isolate which symbol needs this and see why pthread isn't being linked - which test fails to build?

padentomasello replied:
@jacobkahn All tests fails for any library which does not explicitly link to some Threading library, because gtest requires it. (It was getting linked transitively through the audio library before, but now that they are separate, I think we need to add explicitly, https://github.com/flashlight/flashlight/blob/master/flashlight/lib/audio/feature/CMakeLists.txt#L26-L32). Will follow the convention mentioned in the link.

jacobkahn replied:
Sounds good. I've seen this before. That's fine.

jacobkahn commented on flashlight/pkg/vision/CMakeLists.txt:
```diff
@@ -2,7 +2,13 @@ cmake_minimum_required(VERSION 3.10)
-target_link_libraries(fl_pkg_vision PUBLIC flashlight)
+target_link_libraries(fl_pkg_vision PUBLIC flashlight fl_lib_set)
+
+target_include_directories(
```
nit: is this necessary for any of the pkgs? I would've thought this already works as per linking with the `flashlight` target. If it's necessary and I'm wrong, it can be left alone

padentomasello replied:
I was surprised by this as well. Let me double check and get back to you.

jacobkahn commented on flashlight/pkg/text/CMakeLists.txt:
```diff
@@ -3,7 +3,14 @@ cmake_minimum_required(VERSION 3.10)
-  ) 
+  )
+
+target_include_directories(
```
Ditto with the below with this being necessary (?)

jacobkahn commented on flashlight/lib/text/test/CMakeLists.txt:
```diff
@@ -0,0 +1,11 @@
+
```
nit: remove line at beginning of file

jacobkahn commented on flashlight/lib/text/CMakeLists.txt:
```diff
@@ -21,3 +21,9 @@ include(${CMAKE_CURRENT_LIST_DIR}/dictionary/CMakeLists.txt)
+  add_subdirectory(${CMAKE_CURRENT_LIST_DIR}/test)
+endif()
+
+
```
nit: extra line

jacobkahn commented on flashlight/fl/test/CMakeLists.txt:
```diff
@@ -8,7 +8,8 @@ build_test(SRC ${DIR}/common/DynamicBenchmarkTest.cpp LIBS ${LIBS})
 build_test(SRC ${DIR}/common/LoggingTest.cpp LIBS ${LIBS})
 build_test(SRC ${DIR}/common/SerializationTest.cpp LIBS ${LIBS})
-build_test(SRC ${DIR}/common/SequentialBuilderTest.cpp LIBS ${LIBS})
+build_test(SRC ${DIR}/common/SequentialBuilderTest.cpp LIBS ${LIBS}
```
Ditto with my previous comment on the PR where you moved this. I'm slightly impartial to this belonging in `pkg/common`

padentomasello replied:
now in pkg/runtime

jacobkahn commented on .circleci/config.yml:
```diff
@@ -334,9 +370,16 @@ jobs:
           fl_build_app_lm: "OFF"
       - test_flashlight_inside_nvidia_docker:
-          fl_test_apex_dir: "flashlight/lib/test"
+          test_name: "fl_pkg_speech"
```
What's the CI wait time like on this particular job (`ubuntu2004_cuda11_1_docker_flashlight_pkgs`)? If we can parallelize across pkgs, that's well worth it and should cut down on the test time significantly

It'll also make it much easier to reason about which things have broken if one can identify which pkgs they may have broken

padentomasello replied:
The CI time for `ubuntu2004_cuda11_1_docker_flashlight_pkgs` is _just_ over `ubuntu2004_cuda11_1_docker_flashlight_fl_core` ( 26m vs 25 m ).  I agree with the benefits, and actually my first pass on this had each pkg as a separate CI job, but I thought it was a bit cumbersome having so many. Given the very slight time increase, I think I'd still prefer it as is, but let me know if you feel strongly.

jacobkahn replied:
If there's no time increase, that's just fine, and we can leave it as is.

jacobkahn commented on .circleci/config.yml:
```diff
@@ -95,11 +95,14 @@ commands:
     parameters:
+      test_name:
+        type: string
+        default: ""
```
nit: if you leave out the default, it'll be required, which might be what we want

padentomasello replied:
My intention was to not have it be required actually. Let me know if you think it should be, but it's only purpose is to add some information to the circle CI UI, so not really required functionally.

jacobkahn commented on CMakeLists.txt:
```diff
@@ -235,13 +240,59 @@ if (FL_BUILD_CORE)
+
+cmake_dependent_option(FL_BUILD_ALL_PKGS "Set all Flashlight apps ON by default" OFF "FL_BUILD_CORE;FL_BUILD_CONTRIB;" OFF)
+
+# Set build options for pkgs depending on which apps you are build
```
It might make sense to have all of the `cmake_dependent_option` calls for `pkg`s should also have the lib flags in their required options (see the comment on top of this review).

jacobkahn reviewed (commented):
I see we have no options to gate building the lib components - this sort of defeats the purpose since it's all or nothing. Did you intend to add these? We should add them then force them on/off/throw errors (your choice as to behavior) in addition to modifying the `cmake_dependent_option` components.

jacobkahn commented on flashlight/pkg/CMakeLists.txt:
```diff
@@ -1,18 +1,36 @@
+  set(FL_PKG_RUNTIME_DIR ${FL_PKG_DIR}/runtime)
+  include(${FL_PKG_RUNTIME_DIR}/CMakeLists.txt)
+  set(INSTALLABLE_TARGETS ${INSTALLABLE_TARGETS} fl_pkg_runtime)
+  setup_install_headers(${FL_PKG_RUNTIME_DIR} ${FL_INSTALL_INC_DIR_HEADER_LOC}/pkg)
```
Have you checked that the layout of the installed headers is correct given these changes? Probably fine, but worth checking

padentomasello replied:
Checked! Looks good to me.

jacobkahn commented on flashlight/app/CMakeLists.txt:
```diff
@@ -52,15 +32,12 @@ endif ()
 
 if (FL_BUILD_APP_OBJDET)
   message(STATUS "Building flashlight Object Detection app")
-  # TODO (padentomasello) revist if we can pull out some functionality from IMGCLASS
```
👍

jacobkahn reviewed (changes requested):
Submitted early. A few last comments:

Commit:
PR fixes

Commit:
PR fixes

Commit:
Revert "If it's common, probably shouldn't be in ext..."

This reverts commit 3ee51c9ff92ff517a171b61baf84a4dd6c877744.

Commit:
Move ext/common to runtime/common

Commit:
PR fixes

Commit:
PR fixes

Commit:
Remove using namespace from header

Commit:
Revert "namespace ext --> fl"

This reverts commit 0f80173914d526458d988b97a74233ab72cd689d.

Commit:
fl::ext -> fl::pkg::runtime

Commit:
Fix benchmark / pthread inclusion

Commit:
Remove include directories test

Commit:
Trigger CI

Commit:
Add options to gate libraries

Commit:
Remove target include directories from pkgs

Commit:
Automatically set options for libs based on pkgs / apps

Commit:
Move core up in options for cmake dependencies

Commit:
Add pkg build + tests

Commit:
Remove build all pkgs from vcpkg build

Commit:
Small typo fix

Commit:
Fix Halide

Commit:
Make depedencies throw errors on failures

Commit:
Fix CI

Commit:
Fix CI 2

Commit:
Build all pkgs by default

Commit:
Kickoff CI

Commit:
CI Fix

padentomasello commented:
@jacobkahn I went back and forth on what would be best. My initial thought was to automatically set the build flags based on the dependencies. (if FL_PKG_SPEECH is on set FL_LIB_AUDIO to be on since it's required). This actually goes against the current design, which is to use cmake_dependent_option and express the dependency in the other way. (if FL_LIB_AUDIO is not ON, the force FL_PKG_SPEECH to be OFF, even if the user set it to be on). I actually didn't like either of these options because I felt it was confusing to the user. (I set FL_PKG_SPEECH to be on, but it get's set to OFF somewhere in the build system without me knowing. Or I set FL_LIB_SPEECH to OFF and it get's set to ON. either way was somewhat surprising). 

Anyway, long story short, I ended up wrapping cmake_dependent_option with another function (fl_dependent_option), which checks for the proper dependencies and fails if you try to do something that breaks the dependency chain. It's slightly less autonomated, but is very explicit (no surprises). Let me know what you thing.

jacobkahn commented on CMakeLists.txt:
```diff
@@ -52,11 +52,120 @@ option(FL_BUILD_EXPERIMENTAL "Build internal experimental components" OFF)
+
+# Small utility function which wraps cmake_dependent options and throws an error if the user
+# is trying to build something without the proper dependencies.
+macro(FL_DEPENDENT_OPTION OPTION TEXT VAL DEPS FRCE)
```
Would be great if you could do a few quick things to improve this:
1. Move this to `InternalUtils.cmake`. It should be available to all CMake files.
2. Use `function` instead of macro so we can use `cmake_parse_arguments` and named arguments ([example](https://github.com/flashlight/flashlight/blob/c82d4d05e43c21702d57af70c8071282080da1e2/cmake/HalideUtils.cmake#L10-L13), [docs](https://cmake.org/cmake/help/v3.10/command/cmake_parse_arguments.html)). CMake can implicitly cast anything to anything which results in tons of unintended bugs, and having named arguments is a lot clearer. Concretely, after this change, we want the call to look like:
```
fl_dependent_option(
  OPTION FL_BUILD_PKG_SPEECH
  TEXT "Build speech package for Flashlight"
  VAL "${FL_BUILD_ALL_PKGS}"
  DEPS "FL_BUILD_CORE;FL_BUILD_CONTRIB;FL_BUILD_LIB_AUDIO"
  FRCE OFF)
```
Which is much clearer/less error prone.

Optional, but we could also make `DEPS` take multiple things, as in:
```
  DEPS FL_BUILD_CORE FL_BUILD_CONTRIB FL_BUILD_LIB_AUDIO
```
then create the semicolon-separated string as needed for `cmake_dependent_option` so the user doesn't have to deal with the weird formatting. This would basically be
```
set(DEP_STRING "")
for (DEP in ITEMS ${DEPS})
  set(DEP_STRING "${DEP_STRING};${DEP}")
endforeach()
```

padentomasello replied:
1. Will definitely do. Good catch. 
2. The problem with using function is that the scope is not global, so we can't actually perform the check on the variables that are passed in because they only exist on the global scope. (And I think setting the option will also only be for the local scope, but not sure).

jacobkahn commented on CMakeLists.txt:
```diff
@@ -52,11 +52,120 @@ option(FL_BUILD_EXPERIMENTAL "Build internal experimental components" OFF)
+  cmake_dependent_option(${OPTION} ${TEXT} "${VAL}" "${DEPS}" ${FRCE})
+endmacro()
+
+fl_dependent_option(
```
I'd like it if we moved these `fl_dependent_option` calls closer to the targets that they pertain to, sort of like the `cmake_dependent_option` calls we had before. If you move the `fl_dependent_option` macro into `InternalUtils.cmake`, it should be available everywhere.

jacobkahn commented on cmake/TestUtils.cmake:
```diff
@@ -48,6 +49,7 @@ function(build_test)
     PUBLIC
     ${GTEST_IMPORTED_TARGETS}
     ${build_test_LIBS}
+     ${CMAKE_THREAD_LIBS_INIT}
```
👍

jacobkahn commented on .circleci/config.yml:
```diff
@@ -90,16 +93,20 @@ commands:
                 -DFL_BUILD_APP_LM=<< parameters.fl_build_app_lm >> \
-                -DBUILD_SHARED_LIBS=<< parameters.build_shared_libs >> \
+                -DFL_BUILD_ALL_PKGS=<< parameters.fl_build_all_pkgs >> \
+                -DFL_BUILD_SHARED_LIBS=<< parameters.build_shared_libs >> \
```
`BUILD_SHARED_LIBS` is a standard CMake option -- I don't think we want this change
```suggestion
                -DBUILD_SHARED_LIBS=<< parameters.build_shared_libs >> \
```

padentomasello replied:
Ah yes. Good catch. Unintentional.

jacobkahn reviewed (changes requested):
`fl_dependent_option` is a nice abstraction and is a great way of expressing this. I like this a lot more. Just a few final comments with respect to cleanup.

One other thing that I'd appreciate your doing -- updating the documentation/build options and writing up how this works in the [README](https://github.com/flashlight/flashlight/blob/master/README.md) and [Dockerfile things](https://github.com/flashlight/flashlight/tree/master/.docker) if needed.

We also need to properly-gate dependencies in the [`flashlightConfig.cmake` template](https://github.com/flashlight/flashlight/blob/2081246aded0a0e2a7cb9c7fc965424b2af1f7be/cmake/flashlightConfig.cmake.in) based on these new options

Commit:
Move fl_dependent_option to Internal Utils, and options closer to target

Commit:
Opt in

jacobkahn commented on CMakeLists.txt:
```diff
@@ -114,34 +109,42 @@ if (FL_BUILD_TESTS)
+
+fl_dependent_option(FL_BUILD_CORE "Build flashlight core" ON "FL_BUILD_LIB_COMMON" OFF)
+
+# Small utility function which wraps cmake_dependent options and throws an error if the user
```
Nit: remove/move this comment to `InternalUtils.cmake`

jacobkahn commented on cmake/flashlightConfig.cmake.in:
```diff
@@ -96,21 +93,46 @@ endif()
+endif()
+
+# core
+
```
nit: remove newline

jacobkahn commented on flashlight/lib/CMakeLists.txt:
```diff
@@ -1,17 +1,40 @@
 cmake_minimum_required(VERSION 3.10)
 
+# --------------------------- Lib Options  ---------------------------
+# Build all libs by default. In order to selectively build libs,
```
I think this comment needs to be updated?

padentomasello replied:
Yep, good catch.

Commit:
Update README.md

padentomasello commented:
@jacobkahn Everything is opt in now. Documentation is fixed, and fixed dependencies in flashlightConfig.cmake. If these all look good to go, I'll merge it into one branch and import it to phabricactor to start working on some buck targets

jacobkahn commented on README.md:
```diff
@@ -41,7 +41,7 @@ Flashlight is broken down into a few parts:
 - [**`flashlight/fl`**](flashlight/fl) is the core neural network library using the [ArrayFire](https://github.com/arrayfire/arrayfire) tensor library.
 - [**`flashlight/app`**](flashlight/app) are applications of the core library to machine learning across domains.
-- [**`flashlight/ext`**](flashlight/ext) are extensions on top of Flashlight and ArrayFire that are useful across apps.
+- [**`flashlight/pkg**](flashlight/pkg) are packages built on top of Flashlight and ArrayFires for specific purposes (Domain libraries for speech and vision, halide extensions).
```
```suggestion
- [**`flashlight/pkg**](flashlight/pkg) are packages built on top of Flashlight and ArrayFire for specific purposes (Domain libraries for speech, vision, and text, and Halide extensions).
```

jacobkahn commented on README.md:
```diff
@@ -420,12 +486,17 @@ The Flashlight CMake build accepts the following build options (prefixed with `-
+- `flashlight::fl_lib_sequence` — contains flashlight libraries for headers and symbols pertaining to sequences.
+- `flashlight::fl_lib_audio` — contains flashlight libraries for headers and symbols pertaining to audio.
+- `flashlight::fl_lib_text` — contains flashlight libraries for headers and symbols pertaining to text.
 - `flashlight::flashlight` — contains flashlight libraries as well as the flashlight core autograd and neural network library.
```
This is no longer correct (re `flashlight::flashlight`) - can we update it?

padentomasello replied:
We still have the `flashlight` library which is core. We no longer have fl-libraries.

jacobkahn reviewed (commented):
Awesome. Looks great. A few final quick things.

Commit:
Doc fixes

jacobkahn reviewed (approved):
Looks good to me

jacobkahn committed:
Rewrite af-based indexing (#657)

Summary:
Pull Request resolved: https://github.com/flashlight/flashlight/pull/657

Turns out storing an `af::array::array_proxy` as an lvalue isn't really supported and creates lots of problems. Instead, create a visitor which returns an array proxy as a temporary lvalue if needed for in-place operators (assignment, etc) that also returns an array ref.

This is essentially deferred, lazy indexing where indexing occurs after the tensor is used as an rvalue. lvalue tensors that hold proxies can still be operated on as in the case of-in-place assignment.

Also add some explicitly-defined move ctors to avoid some implicit copies that were happening in `Index` and `Tensor`.

Reviewed By: padentomasello

Differential Revision: D29420741

fbshipit-source-id: ebfb2f5e1258afa0a1028957ae4f8d654201bc5d

jacobkahn committed:
Tensor memory interop

Summary:
Functions for interoperating raw memory from the host and device. For now, these capabilities are:
- Construction of tensors from host memory — `fl::Tensor(...)`
- Construction of tensors from device memory — `fl::Tensor(...)`
  - Both of the above using `fl::Location` to specify the location of given memory
- Grabbing scalars from tensors on the host — `fl::Tensor::scalar<T>()` — this only grabs the first value from a tensor if it has > 1 values
- Grabbing a pointer to a tensor on device memory, if applicable — `fl::Tensor::device<T>()`
  - Tensors that have pointers to device memory need to be "unlocked" (this is an implementation requirement since some tensor libraries have this interface) such that their underlying device memory to which the pointer points can be freed.
  - This will become relevant in an implementation of `fl::DevicePtr` that uses `fl::Tensor` to facilitate its RAII-ness.
- Adds `fl::Tensor::shallowCopy()`, also for the purposes of  a `fl::DevicePtr()` implementation later.
- Add constructors for `fl::Tensor` with only `Shape` or `dtype`.

Reviewed By: padentomasello

Differential Revision: D29445385

fbshipit-source-id: 951351a4c493fbd71aa8935c4ffad6dc9025c0be

jacobkahn committed:
Add host memory func to tensor api (#674)

Summary:
Pull Request resolved: https://github.com/flashlight/flashlight/pull/674

Adds a templated function to return a pointer on the host from an `fl::Tensor`.

Will add functions testing some of this memory interop in a future diff once several other prerequisites are in place.

Reviewed By: padentomasello

Differential Revision: D29725738

fbshipit-source-id: f0f5f88077ef532ceafe216a2f70bf28442ff54a

jacobkahn committed:
Remove vcpkg test from CI (#688)

Summary:
See title and https://github.com/microsoft/vcpkg/issues/19022.

Pull Request resolved: https://github.com/flashlight/flashlight/pull/688

Test Plan: CI

Reviewed By: tlikhomanenko

Differential Revision: D29799345

Pulled By: jacobkahn

fbshipit-source-id: 6b511c8680f78458594f0bdc403146b14a996e41

jacobkahn committed:
Tensor where operator (#684)

Summary:
Pull Request resolved: https://github.com/flashlight/flashlight/pull/684

Models [numpy `np.where`](https://numpy.org/doc/stable/reference/generated/numpy.where.html) [without the nonzero behavior]. Uses `af::replace` under the hood.

Reviewed By: benoitsteiner

Differential Revision: D29780654

fbshipit-source-id: 216f79ebc670b7811e75ada349314e7192e52df6

jacobkahn committed:
Move tensor compute/rand functions to TensorBackend interface (#680)

Summary:
Pull Request resolved: https://github.com/flashlight/flashlight/pull/680

See title.

Reviewed By: benoitsteiner

Differential Revision: D29753101

fbshipit-source-id: d091aa4da825f741bb09a61a0db4d62f88c5160f

jacobkahn committed:
Tensor::isContiguous (#675)

Summary:
Pull Request resolved: https://github.com/flashlight/flashlight/pull/675

Adds `isContiguous`. No numpy analog for this, but checking tensor linearity matters for `DevicePtr` and other abstractions.

Reviewed By: benoitsteiner

Differential Revision: D29726689

fbshipit-source-id: b4a772cfe93f11cd0c8acaa46029bc91fed54761

jacobkahn committed:
Fix tensor expression indexing (#683)

Summary:
Pull Request resolved: https://github.com/flashlight/flashlight/pull/683

Expressions-based indexing (where expressions which return per-element bools are used to index into arrays) were indexed in accordance with `af::span`s, which produced unintended results. Only use a single index operator in the case of an expression idx.

Reviewed By: benoitsteiner

Differential Revision: D29780573

fbshipit-source-id: 6333a25591fd4258638d17e14571a333254c8bcb

jacobkahn committed:
Add countNonzero (#677)

Summary:
Pull Request resolved: https://github.com/flashlight/flashlight/pull/677

Add `countNonzero` modeled after [numpy's](https://numpy.org/doc/stable/reference/generated/numpy.count_nonzero.html). Returns per-axis nonzero counts if more than one axis is specified.

Since ArrayFire doesn't have a multi-axis implementation, if more than one axis is specified, operators over that axis, then reduce along the other specified dimensions.

Reviewed By: benoitsteiner

Differential Revision: D29742963

fbshipit-source-id: 6d5d1358a74f23be42fc6cca153137c9ab57a897

jacobkahn committed:
Move binary ops to TensorBackend interface (#679)

Summary:
Pull Request resolved: https://github.com/flashlight/flashlight/pull/679

Move binary operators and tensor creation operators onto the `TensorBackend` API to define implementation requirements for specific backends.

These operators are defined across functions and types for each operator.

Reviewed By: benoitsteiner

Differential Revision: D29748563

fbshipit-source-id: 0569b02d68e3b730fec7a073aba8a30bb5b3bff1

jacobkahn committed:
Fixes to tensor host memory api (#678)

Summary:
Pull Request resolved: https://github.com/flashlight/flashlight/pull/678

Add some fixes to `fl::Tensor::host` and some tests. Add an overload that accepts a type `T*` as an existing buffer and populates that.

Adds:
- `fl::Tensor::size()` (num elements) - replaces `tensor.shape().elements()`
- `fl::Tensor::bytes()` (size in bytes)

Reviewed By: benoitsteiner

Differential Revision: D29744822

fbshipit-source-id: 8ebea91a908b2868aa05695d53253133f0cc9a79

jacobkahn committed:
fl::Tensor matmul (#681)

Summary:
Pull Request resolved: https://github.com/flashlight/flashlight/pull/681

- Title -- adds `matmul`, `matmulTN`, and `matmulNT`.
- Add `fl::Tensor::toVector<T>`
- Add `fl::Tensor::dim(size_t)`

Reviewed By: padentomasello

Differential Revision: D29765124

fbshipit-source-id: 1c92af9e189b3bfeb2912990a5436fdbcee46703

jacobkahn committed:
Tensor any and all (#685)

Summary:
Pull Request resolved: https://github.com/flashlight/flashlight/pull/685

Model `numpy`'s [any](https://numpy.org/doc/stable/reference/generated/numpy.any.html) and [all](https://numpy.org/doc/stable/reference/generated/numpy.all.html).

Reviewed By: benoitsteiner

Differential Revision: D29784833

fbshipit-source-id: 78daf92cc163565c71a24e40b1074b2772f1005b

lunixbochs committed:
Conformer: (breaking change) remove redundant skip connections and fix pointwise conv (#587)

Summary:
**Original Issue**: https://github.com/flashlight/flashlight/issues/582

Please see the linked issue. This PR fixes redundant skip connections in Conformer, as a breaking change. Early tests show this may impact model accuracy by 10% (relative) or more.

This PR also fixes a bug where the output of the first pointwise convolution in a Conformer conv block was unused. The accuracy impact of fixing that bug is not yet known.

I believe as part of merging this PR, the RASR readme should be updated to link to the last flashlight commit where the old RASR Conformer models are expected to work, and new RASR conformer models should be trained and released with the fix.

_The alternative to a breaking change here would be to come up with some kind of constructor hack for the Conformer block to make old models load while applying the fix to new models. That wouldn't be worth it for me though. I'd rather retrain my models to get the improved model accuracy - that's the whole reason I was using a Conformer in the first place._

Pull Request resolved: https://github.com/flashlight/flashlight/pull/587

Reviewed By: tlikhomanenko

Differential Revision: D29891724

Pulled By: jacobkahn

fbshipit-source-id: 1a401502d89813918c9ff9dc7b573eef54e36a3b

lunixbochs committed:
add distributed stub (#614)

Summary:
This makes Windows/Mac builds easier by adding an alternative distributed backend to Gloo.

Closes issue https://github.com/flashlight/flashlight/issues/496

Pull Request resolved: https://github.com/flashlight/flashlight/pull/614

Reviewed By: tlikhomanenko

Differential Revision: D28700959

Pulled By: xuqiantong

fbshipit-source-id: 4792958fe8e4761c7c1facb392ec991ee49e754f

jacobkahn committed:
Tensor pad operator (#690)

Summary:
Pull Request resolved: https://github.com/flashlight/flashlight/pull/690

Add `fl::pad` which mirrors [numpy](https://numpy.org/doc/stable/reference/generated/numpy.pad.html)'s. Only supports `Constant` (with 0), `Edge` and `Symmetric` padding modes for now as those are the only modes ArrayFire supports. More modes can be later as needed.

Reviewed By: benoitsteiner

Differential Revision: D29827519

fbshipit-source-id: c3c221ffebf2b5f9b54aed022bf24c3a8f3921bd

jacobkahn committed:
Add arange, iota (#689)

Summary:
Pull Request resolved: https://github.com/flashlight/flashlight/pull/689

Add `arange` (both a [numpy-like](https://numpy.org/doc/stable/reference/generated/numpy.arange.html) version modeled after ArrayFire's [`range`](https://arrayfire.org/docs/group__data__func__range.htm)) and [`iota`](https://arrayfire.org/docs/group__data__func__iota.htm), which is ArrayFire-specific and has no analog in numpy. There's no numpy-style `arange` equivalent in ArrayFire so the implementation in fl::Tensor uses arithmetic ops to create the same effect.

Reviewed By: benoitsteiner

Differential Revision: D29810629

fbshipit-source-id: da47a0004fcb22e9d44cfc243ff77acfe72f997f

jacobkahn committed:
Rename nDims ndim, add Tensor::isEmpty (#694)

Summary:
Pull Request resolved: https://github.com/flashlight/flashlight/pull/694

See title - renames `Shape::nDims` to be `Shape::ndim` and add `Tensor::ndim()` to adhere to [numpy](https://numpy.org/doc/stable/reference/generated/numpy.ndarray.ndim.html).

Reviewed By: benoitsteiner

Differential Revision: D29888494

fbshipit-source-id: 27f40268cea006901b06f3de2abd494c8dcc8ee1

jacobkahn committed:
Tensor backend conversion - toTensorType (#691)

Summary:
Pull Request resolved: https://github.com/flashlight/flashlight/pull/691

Backend conversion across Tensors.

The implementation requirement for `fl::Tensor::device<T>()` stipulates that memory isn't freed if the underlying tensor falls out of scope unless `unlock` is called — means a device pointer can be used even after the tensor dies since the func takes a `Tensor&&`.

UB can occur if the user creates a shallow tensor copy then converts one of the shallow copy into another tensor (depending on how each backend manipulates the underlying memory), but this could be useful in some cases to save memory, etc, so left as is for the user to control.

Reviewed By: benoitsteiner

Differential Revision: D29856404

fbshipit-source-id: 8062216eebc798c283f697c0dbf8bd1ed5181081

jacobkahn committed:
Add runtime switching of tensor backends (#692)

Summary:
Pull Request resolved: https://github.com/flashlight/flashlight/pull/692

See title.

Since tensor types are parameterized by derived types of `TensorAdapterBase`, the argument to `setDefaultTensorBackend<T>()` is a type. The type gets captured and stored via a closure on `DefaultTensorType` which returns a pointer to the adapter created with the requisite params.

Usage (see test):
```
fl::setDefaultTensorType<ArrayFireTensor>();
// ...
fl::setDefaultTensorType<FooTensor>();
```

Reviewed By: benoitsteiner

Differential Revision: D29869330

fbshipit-source-id: edf573557712d79003e19999cbe6ec4b4d98adb5

jacobkahn committed:
Add funcs to get/set arbitrary data on tensors (#693)

Summary:
Pull Request resolved: https://github.com/flashlight/flashlight/pull/693

Add the ability to get/set arbitrary data on a tensor. Useful for benchmarks/compiler improvements.

Reviewed By: benoitsteiner

Differential Revision: D29886202

fbshipit-source-id: 2c0b3724fa4dd868a1f8feb91b4326dd4b2cc7b1

jacobkahn committed:
fl::Tensor framework readme (#682)

Summary:
Pull Request resolved: https://github.com/flashlight/flashlight/pull/682

See title

Reviewed By: benoitsteiner

Differential Revision: D29275014

fbshipit-source-id: ff530788743de9513f4c52fe442bbfbc9c7cfad8

jacobkahn committed:
Tensor::strides (#695)

Summary:
Pull Request resolved: https://github.com/flashlight/flashlight/pull/695

Add `fl::Tensor::strides`.

Need to investigate whether or not ArrayFire strides dims need to be condensed -- for now, leave leaving 1 dimensions as they are.

Reviewed By: benoitsteiner

Differential Revision: D29891935

fbshipit-source-id: de4624257341999a1a447126a0571c1313b2b1bf

jacobkahn committed:
Add sigmoid and erf Tensor ops (#697)

Summary:
Pull Request resolved: https://github.com/flashlight/flashlight/pull/697

Add sigmoid and [erf](https://en.wikipedia.org/wiki/Error_function) ops to the `fl::Tensor` API for convenience.

Reviewed By: benoitsteiner

Differential Revision: D29893512

fbshipit-source-id: 5b7ff1d6306ec4811de6a7af18b46c0c50134ab7

Commit:
Refactoring SinusoidalPositionEmbedding to avoid concatenation

Summary: Refactoring SinusoidalPositionEmbedding to avoid concatenating vectors, instead using a phase shift to approximate Cosine with Sine.

Reviewed By: tlikhomanenko

Differential Revision: D28733101

fbshipit-source-id: 7089525d3df93e577799c1bb700b1bf581082f1d

jacobkahn committed:
Replace af::eval in fl::eval for array inputs (#705)

Summary:
Pull Request resolved: https://github.com/flashlight/flashlight/pull/705

The `fl::eval(af::array&)` overload had to copy the array into a tensor since it took a reference, which bumped the ArrayFire array refcount and resulted in a copy, causing some OOMs. Use `af::eval` instead to avoid this...

Reviewed By: tlikhomanenko

Differential Revision: D30048654

fbshipit-source-id: 371dadaa8b6f0b9c646a274d2f90781863e0cd61

Commit:
Fixes: 1) decoder in training; 2) transformer set dropout functions (#700)

Summary:
Pull Request resolved: https://github.com/flashlight/flashlight/pull/700

title

Reviewed By: jacobkahn

Differential Revision: D29998888

fbshipit-source-id: ef554ab194c4ab303c40af2d6ddb45ddd02fec40

Commit:
Coverage tests (#706)

Summary:
Pull Request resolved: https://github.com/flashlight/flashlight/pull/706

Cover transformer with tests for changing dropout

Reviewed By: jacobkahn

Differential Revision: D30078735

fbshipit-source-id: 5230d8984ef620cb8d6941ff9215e2b4abf69264

jacobkahn committed:
Switch license to MIT (#710)

Summary:
Pull Request resolved: https://github.com/flashlight/flashlight/pull/710

Swap project licensing to MIT from BSD-3.

Reviewed By: tlikhomanenko

Differential Revision: D30136176

fbshipit-source-id: 1bb3701f752c8ac7637845cf6d30c59b39ba399b

jacobkahn committed:
Tensor API updates pre-autograd integration (#701)

Summary:
Pull Request resolved: https://github.com/flashlight/flashlight/pull/701

- Add a `keepDims` parameter to reductions, which functions like numpy's ([example](https://numpy.org/doc/stable/reference/generated/numpy.amax.html)) which doesn't remoce dimensions that are reduced over in reductions
- Add power overload that takes a double
- Add a `where` overload that takes double(s)
- Fix several edge cases with dimensions not being condensed after reductions
- Fix copy assignment for `Index` and default construction for `range`
- Add a constructor for `Tensor` that takes only an `fl::dtype`
- Remove unused `matmulNT` and `matmulTN` header decls
Make the `ArrayFireBackend` copy/move ctors/assignment operators private as well as the ctor
  - only `ArrayFireBackend::getInstance()` is public ensuring that only one instance can be constructed at a time
- Make `shallowCopy` private and friend with `DevicePtr` (which uses it).
  - This API guarantee means that implementers of dataflow graphs don't need to worry about mutations to things earlier in the graph via shallow copies
  - Tensors passed to `DevicePtr` are never accessible outside of `DevicePtr` and therefore can't introduce additional subtrees into computation graphs

Reviewed By: benoitsteiner

Differential Revision: D29903005

fbshipit-source-id: 37803c6b54ae44456a0ae05fbb16ee7a698b325f

jacobkahn committed:
Upper/lower triangular tensor funcs (#704)

Summary:
Pull Request resolved: https://github.com/flashlight/flashlight/pull/704

See title -- named as per [`tril`](https://numpy.org/doc/stable/reference/generated/numpy.tril.html) and [`triu`](https://numpy.org/doc/stable/reference/generated/numpy.tril.html). Always leave the primary diagonal intact.

Reviewed By: benoitsteiner

Differential Revision: D30022371

fbshipit-source-id: c01ee1c6aad66087a7a2bd7731c328c1983f2269

jacobkahn committed:
Add median tensor func (#702)

Summary:
Pull Request resolved: https://github.com/flashlight/flashlight/pull/702

As titled - see [`median`](https://numpy.org/doc/stable/reference/generated/numpy.median.html).

Reviewed By: benoitsteiner

Differential Revision: D30022372

fbshipit-source-id: 2446ef66f66c55c1ba4fbd14f0738cf8c7f969fe

jacobkahn committed:
Add isinf and sign tensor functions (#703)

Summary:
Pull Request resolved: https://github.com/flashlight/flashlight/pull/703

Modeled after [`isnan`](https://numpy.org/doc/stable/reference/generated/numpy.isnan.html)  - add [`isinf`](https://numpy.org/doc/stable/reference/generated/numpy.isinf.html) - can change naming if others feel strongly.

Adds [`sign`](https://numpy.org/doc/stable/reference/generated/numpy.sign.html) which differs from [ArrayFire's](https://arrayfire.org/docs/group__arith__func__sign.htm) and requires an additional step to zero relevant indices.

Reviewed By: benoitsteiner

Differential Revision: D30022375

fbshipit-source-id: 562de5a36d7c71196f2a7663adaff7a8b50de440
